src/resources/types/surface.c

```diff
@@ -298,7 +300,29 @@ wlc_surface_release(struct wlc_surface *surface)
    struct wlc_surface_event ev = { .surface = surface, .type = WLC_SURFACE_EVENT_DESTROYED };
    wl_signal_emit(&wlc_system_signals()->surface, &ev);
 
-   wlc_handle_release(surface->view);
+   wlc_resource *sub;
+
+   if(!surface->parent) {
+      wlc_handle_release(surface->view);
+   } else {
+
+      struct wlc_surface *parent = convert_from_wlc_resource(surface->parent, "surface");
+      wlc_resource surface_id = convert_to_wlc_resource(surface);
+
+      chck_iter_pool_for_each(&parent->subsurface_list, sub) {
+         if(*sub == surface_id) {
+            chck_iter_pool_remove(&parent->subsurface_list, _I - 1);
+            break;
+         }
+      }
+   }
+
+   chck_iter_pool_for_each(&surface->subsurface_list, sub) {
+       wlc_surface_release(convert_from_wlc_resource(*sub, "surface"));
+   }
```

Comment:
Make this, chck_iter_pool_for_each_call(&surface->subsurface_list, wlc_resource_release_ptr);
Avoid calling *_release for anything that is defined under resources/types/ as these types are tied to the lifetime of wayland resource.